### PR TITLE
Direct Epistemic Uncertainty Predictor

### PR DESCRIPTION
**Pull request by ampudia19**
This branch implements the utilities to perform Bayesian Optimization using deep ensembles but improving on the estimates of uncertainty by accounting for uncertainty linked to the approximation error and not only model variance.

### CODE REVIEW

File: trieste/models/keras/builders.py [hunk 1]
```diff
@@ -80,3 +81,52 @@ def build_vanilla_keras_ensemble(
     keras_ensemble = KerasEnsemble(networks)
 
     return keras_ensemble
+
+
+def build_vanilla_deup(
+    data: Dataset,
+    e_num_hidden_layers: int = 4,
+    e_units: int = 128,
+    e_activation: str = "relu",
```
Comment (clinton0313) [inline trieste/models/keras/builders.py]: This is able to take string or directly the a tf optimizer I believe. tf.keras.optimizers.Optimizer I believe

File: trieste/models/keras/builders.py [hunk 2]
```diff
@@ -80,3 +81,52 @@ def build_vanilla_keras_ensemble(
...
+    f_model_builder: Callable = build_vanilla_keras_ensemble,
+    **f_model_args
+) -> tuple[DeepEnsemble, EpistemicUncertaintyNetwork]:
```
Comment (clinton0313) [inline trieste/models/keras/builders.py:93]: Any TrainableProbablisticModel instead of Deep Ensemble?

File: trieste/models/keras/models.py [hunk 3]
```diff
@@ -367,4 +371,321 @@ def __setstate__(self, state: dict[str, Any]) -> None:
...
+            }
+
+        if self.optimizer.loss is None:
+            self.optimizer.loss = negative_log_likelihood
+
+        self._learning_rate = self.optimizer.optimizer.learning_rate.numpy()
+        
+        model["e_model"].compile(
+            self.optimizer.optimizer,
+            loss="mse",
```
Comment (clinton0313) [inline trieste/models/keras/models.py:463]: Does the model require MSE and is the only possible loss function? I think KerasPredictor already defaults to MSE if not loss function is passed.

File: trieste/models/keras/models.py [hunk 4]
```diff
@@ -367,4 +371,321 @@ def __setstate__(self, state: dict[str, Any]) -> None:
...
+        self._buffer_iter = 1
+        self._data_u = None     
+        self._prior_size = None 
+
+    def __repr__(self) -> str:
+        """"""
+        return f"DirectEpistemicUncertaintyPredictor{self.model!r}, {self.optimizer!r}"
+
+    @property
+    def model(self) -> tuple[DeepEnsemble, tf.keras.Model]:
```
Comment (clinton0313) [inline trieste/models/keras/models.py:477]: Not necessarily DeepEnsemble -> TrainableProbabilistic

File: trieste/models/keras/models.py [hunk 5]
```diff
@@ -367,4 +371,321 @@ def __setstate__(self, state: dict[str, Any]) -> None:
...
+        """
+        Returns two compiled models: A Keras ensemble model and an epistemic uncertainty predictor.
+        """
+        assert issubclass(type(self._model["f_model"]), TrainableProbabilisticModel), (
```
Comment (clinton0313) [inline trieste/models/keras/models.py:481]: Seems unecessary to check this everytime the model is called? At most just when setting the model.

File: trieste/models/keras/models.py [hunk 6]
```diff
@@ -367,4 +371,321 @@ def __setstate__(self, state: dict[str, Any]) -> None:
...
+        :return: The predicted mean and variance of the observations at the specified
+            ``query_points``.
+        """
+        if not tf.is_tensor(query_points):
+            query_points = tf.convert_to_tensor(query_points)
+        if query_points.shape.rank == 1:
+            query_points = tf.expand_dims(query_points, axis=-1)
+
+        f_pred, f_var = self.model[0].predict(query_points)
+        f_pred, f_var = tf.cast(f_pred, dtype=tf.float64), tf.cast(f_var, dtype=tf.float64)
```
Comment (clinton0313) [inline trieste/models/keras/models.py]: Is the model outputting something other than float64 sometimes? It should always output whatever dtype was inferred from building the architecture off of the initial data no? I think this is the case for the f_predictors since they are all the other trainable probabilistic models out there.

File: trieste/models/keras/models.py [hunk 7]
```diff
@@ -367,4 +371,321 @@ def __setstate__(self, state: dict[str, Any]) -> None:
...
+        
+        model["e_model"].compile(
+            self.optimizer.optimizer,
+            loss="mse",
+            metrics=[self.optimizer.metrics],
+        )
+
+        self._model = model
+        self._init_buffer = init_buffer
+        self._buffer_iter = 1
```
Comment (clinton0313) [inline trieste/models/keras/models.py]: Shouldn't this be a hyperparameter?

File: trieste/models/keras/models.py [hunk 8]
```diff
@@ -367,4 +371,321 @@ def __setstate__(self, state: dict[str, Any]) -> None:
...
+        the main proposed algorithm in <cite data-cite="jain2022"/>.
+
+        :param dataset: The data with which to optimize the model.
+        """     
+        # optional init buffer
+        if self._data_u is None:
+            self.density_estimator = KernelDensityEstimator(kernel="gaussian")
+            if self._init_buffer:
+                if dataset.query_points.shape[0] > 1:
+                    self._data_u = self.uncertainty_buffer(dataset=dataset, iterations=self._buffer_iter)
```
Comment (clinton0313) [inline trieste/models/keras/models.py]: seems unecessary to have iterations argument since it is a class variable anyways. Can just call self._buffer_iter inside the method.

File: trieste/models/keras/models.py [hunk 9]
```diff
@@ -367,4 +371,321 @@ def __setstate__(self, state: dict[str, Any]) -> None:
...
+                    self._prior_size = dataset.query_points.shape[0]
+                else:
+                    self._init_buffer = False
+            if self._data_u is None:
+                self._data_u = Dataset(
+                    tf.zeros([0, dataset.query_points.shape[-1] + 2], dtype=tf.float64),
```
Comment (clinton0313) [inline trieste/models/keras/models.py]: Is this dtype=float64 necessary? I think the backend is already set to float64 typically. If it actually stops the model from working I think it would be better to infer the dtype (same way already done for the architecture) from the dataset and could make a self._dtype, but I think this is not necessary. Would cause problems if ever worked in something other than float64.

File: trieste/models/keras/models.py [hunk 10]
```diff
@@ -367,4 +371,321 @@ def __setstate__(self, state: dict[str, Any]) -> None:
...
+        data = tf.concat((dataset.query_points, dataset.observations), axis=1)
+        for _ in tf.range(iterations):
+            for set in tf.random.shuffle(tf.split(data, 2, axis=0)):
+                data_ = Dataset(set[:,:-1], tf.expand_dims(set[:,-1], axis=1))
+
+                # stationarizing feature: variance
+                f_ = self.__copy__().model[0]
+                f_.optimize(data_)
+                f_pred, f_var = f_.predict(dataset.query_points)
+                f_pred, f_var = tf.cast(f_pred, dtype=tf.float64), tf.cast(f_var, dtype=tf.float64)
```
Comment (clinton0313) [inline trieste/models/keras/models.py]: same dtype thing

File: trieste/models/keras/models.py [hunk 11]
```diff
@@ -367,4 +371,321 @@ def __setstate__(self, state: dict[str, Any]) -> None:
...
+        :return: A new dataset for the uncertainty prediction, which includes the previous
+            observations and the new batch of points.
+
+        """
+        new_points = query_points[self._prior_size:,:]
+        new_observations = observations[self._prior_size:,:]
+
+        # stationarizing feature: variance
+        f_pred, f_var = self.model[0].predict(new_points)
+        f_pred, f_var = tf.cast(f_pred, dtype=tf.float64), tf.cast(f_var, dtype=tf.float64)
```
Comment (clinton0313) [inline trieste/models/keras/models.py]: same dtype thing

File: trieste/models/keras/models.py [hunk 12]
```diff
@@ -367,4 +371,321 @@ def __setstate__(self, state: dict[str, Any]) -> None:
...
+
+        # stationarizing feature: density
+        density_scores = self.density_estimator.score_samples(new_points)
+
+        new_data_u = Dataset(
+            tf.concat((new_points, f_var, density_scores), axis=1),   # [N, D+2]
+            tf.pow(tf.subtract(f_pred, new_observations), 2)
         )
+        return Dataset(
```
Comment (clinton0313) [inline trieste/models/keras/models.py]: I think you can just add datasets. self._data_u + new_data_u

File: trieste/models/keras/models.py [hunk 13]
```diff
@@ -367,4 +371,321 @@ def __setstate__(self, state: dict[str, Any]) -> None:
...
+        over two copies of the dataset's observations to account for the
+        stationarizing features and the two targets: the target outcome for the main
+        predictor and the squared loss for the auxiliary predictor. The procedure follows
+        the main proposed algorithm in <cite data-cite="jain2022"/>.
+
+        :param dataset: The data with which to optimize the model.
+        """     
+        # optional init buffer
+        if self._data_u is None:
+            self.density_estimator = KernelDensityEstimator(kernel="gaussian")
```
Comment (clinton0313) [inline trieste/models/keras/models.py:539]: Should kernel type be a hyperparameter? if not why even bother with any argument at all?

File: trieste/models/keras/utils.py [hunk 14]
```diff
@@ -98,3 +102,62 @@ def negative_log_likelihood(
...
+
+        :param query_points: The points to use while searching the optimal bandwidth.
+        :param bandwidth: A fixed bandwidth can be instead provides, by default is None.
+        """
+        if bandwidth is not None:
+            self.bandwidth = bandwidth
+        else:
+            if query_points.shape[0] > 1:
+                bandwidth_search_space = np.logspace(-3,1,50)
+                grid = GridSearchCV(
```
Comment (clinton0313) [inline trieste/models/keras/utils.py]: Maybe use randomizedCV or at least an option. Can really speed things up for us.

Comment (clinton0313) [reply trieste/models/keras/utils.py]: Maybe should just have another builder for the kernel density. potentially CV for 50 different bandwidths is too expensive and might not be doing too much? also can maybe get away with fewer cross validate folds? I feel like these plus the potential use of RandomizedCV are very relevant parameters especially for our near future of benchmarking, since I remember you saying this CV is a huge time sink. I think it warrants a separate builder, or extra functionality to current builder to do this. I.e. a builder that has args: bandwidth searchspace, type of search, type of kernel,

**Review by clinton0313 — COMMENTED** [review]
Aside from the dtype and kernel density estimator thing, maybe also leave out the mega basic_DEUP_tests.ipynb. I don't even want to know about 804,665 additions haha..

File: tests/unit/models/keras/test_models.py [hunk 15]
```diff
@@ -50,6 +57,11 @@ def _bootstrap_data_fixture(request: Any) -> bool:
     return request.param
 
 
+@pytest.fixture(name="init_buffer", params=[False, True])
+def _init_buffer_fixture(request: Any) -> bool:
+    return request.param
+
+
 def _get_example_data(query_point_shape: ShapeLike) -> Dataset:
     qp = tf.random.uniform(tf.TensorShape(query_point_shape), dtype=tf.float64)
```
Comment (clinton0313) [inline tests/unit/models/keras/test_models.py:66]: Maybe the dtype issue?

File: trieste/models/keras/builders.py (hunk 1 above)
Comment (ampudia19) [reply trieste/models/keras/builders.py]: I guess you mean `tf.keras.layers.Activation`?

File: trieste/models/keras/builders.py (hunk 2 above)
Comment (ampudia19) [reply trieste/models/keras/builders.py:93]: That is the idea for when we perhaps work to integrate these models on trieste. As is, for now it will only work with deep ensembles.

File: trieste/models/keras/models.py (hunk 3 above)
Comment (ampudia19) [reply trieste/models/keras/models.py:463]: This took me a while to wrap my head around, but KerasOptimizer is running with negative_log_likelihood (as it uses deep ensembles as its main predictor, and the theory part of the DE paper suggests using NLL). Hence, I must "override" the loss for the epistemic uncertainty model, which does not require MSE but is I guess the most obvious alternative.

File: trieste/models/keras/models.py (hunk 4 above)
Comment (ampudia19) [reply trieste/models/keras/models.py:477]: Not for now (will only work with Deep Ensemble), and having DeepEnsemble is more informative. I guess the confusion is from the asserts (which I should take out as per Hrvoje recommendations) where I check if subclass is for TrainableProbabilistic.

File: trieste/models/keras/models.py (hunk 5 above)
Comment (ampudia19) [reply trieste/models/keras/models.py:481]: I should just take it out, agreed. It also seems costless to check, tho.

File: trieste/models/keras/models.py (hunk 6 above)
Comment (ampudia19) [reply trieste/models/keras/models.py]: Surprisingly, no. The data in _get_example_data is float64, and the output predictions (as standalone or as main predictor here) of the deep ensembles are float32. Since they are not concatenated this is not flagged as an issue, but with the tf.concats here I need to homogenize dtypes.

File: trieste/models/keras/models.py (hunk 7 above)
Comment (ampudia19) [reply trieste/models/keras/models.py]: In fact, the idea would be to have a single integer parameter: 0 (default) is no initial buffer, otherwise any number is the number of full loops. I didn't change it out of just how much I have used the init_buffer param everywhere (I added it y'day to include it as a models_test), but you make a good point.

File: trieste/models/keras/models.py (hunk 8 above)
Comment (ampudia19) [reply trieste/models/keras/models.py]: A very good point!

File: trieste/models/keras/models.py (hunk 9 above)
Comment (ampudia19) [reply trieste/models/keras/models.py]: The point is similar to the one above: data points, variance and density must all have the same dtype to concatenate them and use them in the epistemic predictor. Any solution should still guarantee all these tensors share a dtype.

File: trieste/models/keras/models.py (hunk 7 above)
Comment (ampudia19) [reply trieste/models/keras/models.py]: I changed it.

File: trieste/models/keras/models.py (hunk 10 above)
Comment (ampudia19) [reply trieste/models/keras/models.py]: Same answer, unfortunately :(

File: trieste/models/keras/models.py (hunk 11 above)
Comment (ampudia19) [reply trieste/models/keras/models.py]: Same :(

File: trieste/models/keras/models.py (hunk 12 above)
Comment (ampudia19) [reply trieste/models/keras/models.py]: Yes!

File: trieste/models/keras/models.py (hunk 13 above)
Comment (ampudia19) [reply trieste/models/keras/models.py:539]: Same with DeepEnsembles, to facilitate / remind me that this should eventually be a hyperparameter. But indeed no HP at the moment, so I'll simplify it.

File: trieste/models/keras/utils.py (hunk 14 above)
Comment (ampudia19) [reply trieste/models/keras/utils.py]: No, this CV is not a time sink, it's the init_buffer (which does CV on the ensembles, optimizes, predicts, etc. to create the initial dataset for the e_estimator). This is actually very very fast, we don't have many points so leave-one-out CV is feasible. It also follows the authors proposition.

Comment (ampudia19) [reply trieste/models/keras/utils.py]: And as per your other comment above, I've changed it now so that we can also benchmark on CV loops of init_buffer (CV is set to 2 tho, otherwise computational cost just increases too much)

File: tests/unit/models/keras/test_models.py (hunk 15 above)
Comment (ampudia19) [reply tests/unit/models/keras/test_models.py:66]: Good point

File: trieste/models/keras/models.py (hunk 6 above)
Comment (ampudia19) [reply trieste/models/keras/models.py]: Ok, so I did a general fix. Now I infer the data dtype, and ensure variance and density use the same (this required quite a few changes in the density class, but it works). I included a test that validates the changes. Thanks for bringing it up :)

File: trieste/models/keras/builders.py (hunk 1 above)
Comment (clinton0313) [reply trieste/models/keras/builders.py]: Yes
